GPUP: ImageBuffer::putPixelBuffer spends time copying data at GPUP side

kkinnunen-apple opened the pull request:
#### b145f46da3ba2b6d9d51afa7ef0bbcbfa2f585c3
<pre>
GPUP: ImageBuffer::putPixelBuffer spends time copying data at GPUP side
<a href="https://bugs.webkit.org/show_bug.cgi?id=290298">https://bugs.webkit.org/show_bug.cgi?id=290298</a>
<a href="https://rdar.apple.com/147725500">rdar://147725500</a>

Reviewed by Simon Fraser.

ImageBuffer::putPixelBuffer(PixelBuffer&amp;) would require that the
caller holds a real PixelBuffer. Holding such a object means that the
object owns the data and that the data is mutable. This meant that
RemoteImageBuffer had to copy the data from the IPC buffers.

Fix by making the API more narrow in form of
ImageBuffer::putImageData(PixelBufferSourceView).

PixelBufferSourceView can reference the buffer data directly for the
duration of the call and convey that the data is immutable.

Theoretically ImageBuffer::putPixelBuffer(SomethingReffable&amp;) could
support API that would allow for more efficient data passing, in form
of a pattern:
  Ref passablePixelBuffer = imageBuffer-&gt;allocatePixelBuffer(...);
  fillPixelBuffer(passablePixelBuffer, ...);
  imageBuffer-&gt;putPixelBuffer(passablePixelBuffer, ...);

However, it wasn&apos;t implemented and in pratice it is not that realistic.
Primarily, the web-facing APIs cannot use the right allocate at the time
the allocation happens from the web side (new ImageData(..)). The
putPixelBuffer API as it is today couldn&apos;t store a reference and read it
later, since the reference was of mutable PixelBuffer.

If the above pattern is needed for non-web facing APIs (filter usage),
a distinct API call that really supports it can be added separately.
The lowest level implemenation of such a API will still use the
PixelBufferSourceView added here.

* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createCompletionHandler):
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::PageConsoleClient::screenshot):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::putPixelBuffer):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::putPixelBuffer):
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/ImageBufferDisplayListBackend.h:
* Source/WebCore/platform/graphics/NullImageBufferBackend.cpp:
(WebCore::NullImageBufferBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/NullImageBufferBackend.h:
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBufferSourceView::PixelBufferSourceView):
(WebCore::PixelBufferSourceView::create):
(WebCore::PixelBufferSourceView::format const):
(WebCore::PixelBufferSourceView::size const):
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.cpp:
(WebCore::ImageBufferCairoSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::imageBufferFromPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp:
(WebCore::ImageBufferSkiaUnacceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::putPixelBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
(WebKit::DynamicContentScalingImageBufferBackend::putPixelBuffer):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::putPixelBufferForImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: <a href="https://commits.webkit.org/292641@main">https://commits.webkit.org/292641@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/680223a97d25987300911f42f4c193cf756fd2f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6402 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101737 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47184 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98709 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16574 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24715 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73672 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30889 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99667 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12474 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87420 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54007 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/96151 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12232 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5220 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46512 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82334 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5308 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103760 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23732 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17302 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82724 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23982 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83473 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82107 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20608 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26752 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4281 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17205 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23694 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28849 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23353 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26833 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25094 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/PixelBuffer.h:
```diff
@@ -77,4 +79,36 @@ class PixelBuffer : public ThreadSafeRefCounted<PixelBuffer> {
     std::span<uint8_t> m_bytes;
 };
 
+// Type to use for functions that use the PixelBuffer data as source during the call, but do not store a reference to the object or modify the data.
```
Not necessarily a problem, but this very close to the same type as ConstPixelBufferConversionView.  Might make sense to unify them, but I won't push it.

This is also quite a bit closer to the original design of PixelBuffer when I added it a few years ago. I looks like in be24c392b5e0 it got turned into a ref-counted type.

Also, unless types are relatively trivial, we tend to prefer a file per type, so I think this would be better suited to its own file.

kkinnunen-apple replied:
>ConstPixelBufferConversionView. Might make sense to unify them, but I won't push it.

I considered it, but prefer not in this patch. Definitively later 
There's the stride field in ConstPixelBufferConversionView that would then need special attention (as in need to get it right in IPC).

kkinnunen-apple replied:
> Also, unless types are relatively trivial, we tend to prefer a file per type, so I think this would be better suited to its own file.

I'll punt this one on the count of: 
- The patch already being written and changing this part is quite much work that often breaks with platforms I cannot build
- The primary feature of the view is that PixelBuffer& gets converted to the view automatically. This would need the caller to include PixelBufferSourceView.h when seemingly they don't use it
- It's relatively trivial type for some subjective view
- It uses the internals of PixelBuffer, so maybe it's easier to maintain, for now, in the same file
- There's no significant downside to having it in the same file

I'd r+ a patch moving it to its own header, though.

weinig commented on Source/WebCore/platform/graphics/PixelBuffer.h:
```diff
@@ -77,4 +79,36 @@ class PixelBuffer : public ThreadSafeRefCounted<PixelBuffer> {
+    const PixelBufferFormat& format() const { return m_format; }
+    IntSize size() const { return m_size; }
+    std::span<const uint8_t> bytes() const LIFETIME_BOUND { return m_bytes; }
+private:
```
Missing newline.

weinig commented on Source/WebCore/platform/graphics/PixelBuffer.h:
```diff
@@ -77,4 +79,36 @@ class PixelBuffer : public ThreadSafeRefCounted<PixelBuffer> {
+        : PixelBufferSourceView(pixelBuffer.format(), pixelBuffer.size(), pixelBuffer.bytes())
+    {
+    }
+    static std::optional<PixelBufferSourceView> create(const PixelBufferFormat& format, const IntSize& size, std::span<const uint8_t> bytes)
```
Is this used anywhere?

kkinnunen-apple replied:
It's the IPC constructor that validates.

kkinnunen-apple committed:
GPUP: ImageBuffer::putPixelBuffer spends time copying data at GPUP side
https://bugs.webkit.org/show_bug.cgi?id=290298
rdar://147725500

Reviewed by Simon Fraser.

ImageBuffer::putPixelBuffer(PixelBuffer&) would require that the
caller holds a real PixelBuffer. Holding such a object means that the
object owns the data and that the data is mutable. This meant that
RemoteImageBuffer had to copy the data from the IPC buffers.

Fix by making the API more narrow in form of
ImageBuffer::putImageData(PixelBufferSourceView).

PixelBufferSourceView can reference the buffer data directly for the
duration of the call and convey that the data is immutable.

Theoretically ImageBuffer::putPixelBuffer(SomethingReffable&) could
support API that would allow for more efficient data passing, in form
of a pattern:
  Ref passablePixelBuffer = imageBuffer->allocatePixelBuffer(...);
  fillPixelBuffer(passablePixelBuffer, ...);
  imageBuffer->putPixelBuffer(passablePixelBuffer, ...);

However, it wasn't implemented and in pratice it is not that realistic.
Primarily, the web-facing APIs cannot use the right allocate at the time
the allocation happens from the web side (new ImageData(..)). The
putPixelBuffer API as it is today couldn't store a reference and read it
later, since the reference was of mutable PixelBuffer.

If the above pattern is needed for non-web facing APIs (filter usage),
a distinct API call that really supports it can be added separately.
The lowest level implemenation of such a API will still use the
PixelBufferSourceView added here.

* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createCompletionHandler):
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::PageConsoleClient::screenshot):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::putPixelBuffer):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::putPixelBuffer):
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/ImageBufferDisplayListBackend.h:
* Source/WebCore/platform/graphics/NullImageBufferBackend.cpp:
(WebCore::NullImageBufferBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/NullImageBufferBackend.h:
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBufferSourceView::PixelBufferSourceView):
(WebCore::PixelBufferSourceView::create):
(WebCore::PixelBufferSourceView::format const):
(WebCore::PixelBufferSourceView::size const):
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.cpp:
(WebCore::ImageBufferCairoSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::imageBufferFromPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp:
(WebCore::ImageBufferSkiaUnacceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::putPixelBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
(WebKit::DynamicContentScalingImageBufferBackend::putPixelBuffer):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::putPixelBufferForImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::putPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: https://commits.webkit.org/292641@main
